libyul/YulStack.h

```diff
@@ -152,6 +152,13 @@ class YulStack: public langutil::CharStreamProvider
 
 	void optimize(yul::Object& _object, bool _isCreation);
 
+	/// Prints the Yul object stored internally and parses it again.
+	/// This ensures that the debug info in the AST matches the source that printing would produce
+	/// rather than the initial source.
+	/// @warning Does not update the error list or the original source (@a m_charStream) to make
+	/// it possible to report errors to the user even after the optimization has been performed.
+	void reparse();
```

Comment:
Sounds like something that could potentially also be solved by keeping the original parsing result / object around and the `optimize` method create a new, optimized version of it (perhaps in the spirit of the immutable AST PR). Then, I imagine, there also wouldn't be this mix of properties where some refer to data before optimization, some to data after optimization and you could potentially get rid of the `YulStack` in `YulStack` implementation of `reparse`.

Reply:
True, that would be a much better design, more similar to `CompilerStack`. `YulStack` is just wildly inconsistent. Parsing is stateful but creates new things (stores parsing result), then optimization is stateful but modifies things in place, and then assembling is stateless and not in-place (just returns the result to the caller). It should be more like a proper pipeline with each stage filling its artifact based on the result of previous stages.

Still, I'm not to keen to get into that refactor right now. I'd get rid of `YulStack` from `reparse()` but I would still basically have to do the refactor I avoided, i.e. create stateless helpers for parsing and analysis so that I can run them on different objects. I wanted this to be a quick fix so that I get the cache to PR to a mergeable state before I'm off tomorrow.

So, if there are no bugs or bigger issues here, I'd rather merge this as is to unblock the caching PR. We can always do that refactor on top of them, later.